Parse and import both rpm and deb packages metadata

waterflow80 opened the pull request:
## What does this PR change?

Parse both rpm and deb packages/repositories' metadata files and import the data to the database. 


## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- No tests: still in progress

- [x] **DONE**

## Links

Issue(s): #
Port(s): # **add downstream PR(s), if any**

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

agraul committed:
Create lzreposync subdirectory

lzreposync will be a spacewalk-repo-sync replacement written in Python.
It uses a src layout and a pyproject.toml. The target Python version is
3.11, compatibility with older Python versions is explicitly not a goal.

waterflow80 committed:
Correct error message

waterflow80 committed:
Add remote_path column

Added the remote_path column that will hold the remote path/
url of a given package.

This information will help locate the package later-on on the
remote repository and download it.

waterflow80 committed:
Add expand_full_filelist parameter

A boolean argument that checks whether we should call the
header.hdr.fullFilelist()

We added this argument to disable the header.hdr.fullFilelist()
function only for the lzreposync service.

waterflow80 committed:
Update deprecated method

The inspect.getargspec() method is deprecated in Python 3
It can be replaced by inspect.getfullargspec()

waterflow80 committed:
Add import_signatures parameter

The import_signatures is a boolean argument that specifies
whether we should execute the _import_signatures() method.

We added this parameter to disable the _import_signatures()
method for the lzreposync service.

waterflow80 committed:
Implement Primary.xml file parser

Parsing the rpm's Primary.xml packages metadata file using
pulldom xml parser as a memory efficient parsing library.

Note that some attributes in the returned parsed object are
faked, and maybe filled in elsewhere.

The faking of some of the data is done because some
attributes are required by the importer service.

waterflow80 committed:
Implement filelists.xml file parser

Parsing the rpm's filelists.xml metadata file using
pulldom xml parser as a memory efficient parsing library.

The parser parses the given filelists.xml file (normally in gz
format), and cache the filelist information of each package
in a separate file in the cache directory, using the package's
hash as the filename, with no file extension.

waterflow80 committed:
Implement full rpm metadata parsing

Using both primary_parser and filelists_parser, return the full
packages' metadata, pacakge by package, using lazing parsing.

Note that there some attributes that are faked, because we
can't fetch them now, and they're required by the package
importer later-on.
However, we can fake them more efficiently, using less memory.

waterflow80 committed:
Parse and import rpm patches/updates

Parsed the update-info.xml file and imported the parsed
patches/updates to the database.

We used pretty much the same code from the old Reposync class.

waterflow80 committed:
Import parsed rpm & deb packages to db

Import the parsed rpm and debian packages to the database in
batche, and associate each pacakge with the corresponding
channel

waterflow80 committed:
Implement the deb Packages md file

Parsed the debian Packages metadata file in a lazy way and
yield the metadata of each package separately.

waterflow80 committed:
Implement the Translation file parser

Parsed the debian's Translation file that contains the full
description of packages, grouped by description-md5, and
cache the parsed descriptions in a cache directory.

waterflow80 committed:
Implement full deb metadata parsing

Using both packages_parser and translation_parser, return the
full packages' metadata, pacakge by package, using lazing
parsing

Also set the debian repository's information in a DebRepo
class

waterflow80 committed:
Fetch repository information from the db

Given the channel label, fetch important repository's
information form the database, and store it in a temporary
object RepoDTO

waterflow80 committed:
Complete lzreposync service entry point

Added the necessary command line arguments.

Identify the target repositories, prepare the datastructures,
and execute the lazy synchronization of repositories/packages.

waterflow80 committed:
Add new dependency

Added a new dependency python-gnupg used to verify repo
signature.

waterflow80 committed:
Add unit tests for rpm metadata parsers

waterflow80 committed:
Delete no longer used files

waterflow80 committed:
Remove already defined function

waterflow80 committed:
Fix linting complain

Ignored two linting complains about rasing exceptions floowing the
approach in the old reposync.

We can enhance the code instead of doing this though.

waterflow80 committed:
Complete code for lzreposync version 0.1

This commit completes almost all the logic and use cases
of the new lazy reposync.

**Note** that this commit will be restructured and possibly
divided into smaller and more convenient commits.
This commit is for review purposes.

waterflow80 committed:
Complete tests for lzreposync service

waterflow80 committed:
Fix error: too many clients already

Seemingly this error happened because we reached the maximum number
of unclosed db connections. And thought that this might be due to
the fact that the close() method in the Database class was not
implemented, and the rhnSQL.closeDB() was not closing any connection.

However, we're still hesitating about whether this is the root cause
of the problem, because the old(current) reposync is was using it
without any error.

waterflow80 committed:
Complete latest version

This is the latest and almost the final version of the
lzreposync service. (gpg sig check not complete)

It contains pretty much all the necessary tests,
including the ones for updates/patches import.

Some of the remaining 'todos' are either for code
enhancements or some unclear concepts that
will be discussed with the team.

Of course, this commit will be split into smaller
ones later after rebase.

aaannz commented on schema/spacewalk/common/tables/rhnPackage.sql:
```diff
@@ -56,6 +56,7 @@ CREATE TABLE rhnPackage
                          CONSTRAINT rhn_package_compat_check
                              CHECK (compat in ( 1 , 0 )),
     path             VARCHAR(1000),
+    remote_path	     VARCHAR(1000),
```
Is there a reason, except 'we did it like that before', to limit path to 1000 characters?
In postgresql varchar is implemented using TEXT datatype with string length check on top of it. So using limited VARCHAR is actually a tiny little slower then just using TEXT and has no space or indexing benefit.

So if there is no reason to limit this to 1000 characters, please use TEXT as a datatype here.

waterflow80 replied:
Absolutely..thanks for the review,
I did not have any specific reason for using `VARCHAR(1000)`...I'll change it to `TEXT` instead.

agraul commented on python/lzreposync/src/lzreposync/__init__.py:
```diff
@@ -0,0 +1,218 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(
```
In the future, i.e. not in this PR, we should align the CLI interface as much as possible with `spacewalk-repo-sync`. We don't have to support all the flags, a subset is fine. Right now we have conflicting flags (e.g. `-d` means _debug_ here but _dry-run_ in `spacewalk-repo-sync`)

agraul commented on python/lzreposync/src/lzreposync/__init__.py:
```diff
@@ -0,0 +1,218 @@
+
+    arch = args.arch
+    if arch != ".*":
+        # pylint: disable-next=consider-using-f-string
+        arch = "(noarch|{})".format(args.arch)
```
Instead of silencing pylint, we should follow the rules we have set it for it. In this case it's as simple as:
```suggestion
        arch = f"(noarch|{args.arch})"
```

agraul commented on python/lzreposync/src/lzreposync/deb_metadata_parser.py:
```diff
@@ -0,0 +1,70 @@
+
+from lzreposync.packages_parser import PackagesParser
+from lzreposync.translation_parser import TranslationParser
+from spacewalk.satellite_tools.syncLib import log, log2
```
Just a heads-up: logging in reposync currently suffers from a split-brain problem. We have `rhnLog` (with `syncLib` defining wrapper functions) one the one hand and `logging` on the other. The latter is also used by 3rd-party libraries (e.g. `requests`). It's one of the things on our list to clean up, so it does not matter much what you use now. That said, I suggest to stick to one of them to make it easy for yourself.

agraul commented on python/lzreposync/src/lzreposync/deb_metadata_parser.py:
```diff
@@ -0,0 +1,70 @@
+
+
+#  pylint: disable-next=missing-class-docstring
+class DEBMetadataParser:
```
This class only has one method, I would move that to a "free" function at the module level. Adding another layer makes the name longer but does not give us any benefit

waterflow80 replied:
I wanted to add some abstraction when using OOP...
For example in function [import_packages_in_batch_from_parser()](https://github.com/uyuni-project/uyuni/blob/8a49313f35c903cb79926e6a5a3bcb7264b43612/python/lzreposync/src/lzreposync/import_utils.py#L89) (used in the tests) we're making use of this abstraction to import packages using abstract `parser` objects that implement the `parse_packages_metadata()` method.

And I'm afraid that using "free functions might add some complexity especially when passing a function as a parameter to another function instead of an object, in the above scenario.    

However I'm not sure whether it's the best way to approach things in our case.
(`pylint` also complains about using one method within a class `R0903: Too few public methods (1/2)`)

agraul replied:
To use `import_packages_in_batch_from_parser`, you first instantiate a parser object. The parser object is only used to then call the single public method it has. In other words, at the call time of `import_packages_in_batch_from_parser`, the caller has to decide which parser to pass (rpm or deb). Therefore the caller could also pass specific functions (e.g. `parse_rpm_metadata()` or `parse_deb_metadata()`) to `import_packages_in_batch_from_parser`, or call these generator functions themselves and pass the generator object instead.

In simplified/pseudo code:
```python
# parsing code
def parse_rpm_metadata(primary_xml, repository_url, arch_filter, filelists_xml, cache_dir):
    primary_parser = PrimaryParser(primary_xml, repository_url, arch_filter)
    filelists_parser = FilelistsParser(filelists_xml, cache_dir, arch_filter)
    filelitsts_parser.parse_filelists()
    for package in primary_parser.parse():
        ...
        yield rpm_package

# import code
def import_in_batches(packages, batch_size, channel=None, compatible_archs=None):
    total = 0
    for i, batch in enumerate(batched(packages, batch_size)):
        failed, *_ = import_package_batch(
            to_process=batch,
            compatible_archs=compatible_archs,
            batch_index=i,
            to_link=True,
            channel=channel
        )
        total += failed
    return total

# test code
def test_parse_and_import():
    packages = parse_rpm_metadata(primary, repo_url, arch_filter, filelists)
    failed = import_in_batches(packages, 20)
    assert ...
```

I like this more than passing a function because we decouple the different concerns. Import code shouldn't need to bother with getting the data to import, that's someone else's job.
That said, in Python it's not a problem to pass functions around, they are first-class objects, e.g.:
```python
>>> import random
>>> def foo():
...     print("Hello, I'm Foo")
... 
>>> def maybe(f):
...     if random.randint(0, 1) == 1:
...        f() 
...     else:
...        print("No luck. Maybe next time")
... 
>>> for _ in range(10):
...     maybe(foo)
... 
Hello, I'm Foo
No luck. Maybe next time
No luck. Maybe next time
No luck. Maybe next time
No luck. Maybe next time
Hello, I'm Foo
No luck. Maybe next time
Hello, I'm Foo
Hello, I'm Foo
No luck. Maybe next time

waterflow80 replied:
Very good explanation..and the approach is much better too.
Thanks.

agraul commented on python/lzreposync/src/lzreposync/deb_repo.py:
```diff
@@ -0,0 +1,236 @@
+
+
+#  pylint: disable-next=missing-class-docstring
+class DebRepo(Repo):
```
We should consolidate the debian repo parsing. Right now it's split into `repo_plugins.deb_src`, `common.repo` and this `deb_repo` (which appears to be a copy of the class in `deb_src`

agraul reviewed (commented):
This is the first part, I will continue but want to send some comments already. Besides what I wrote, I suggest to go over the `pylint` comments and try to resolve the complaints instead of disabling the checks. `lzreposync` is a new program and should meet our current standards better than the old code :sweat_smile:

waterflow80 committed:
Optimize code and do some cleanup

- Removed some todos.
- Changed some sql queries with equivalent ones using
JOIN...ON.
- Some other minor cleanup

waterflow80 committed:
Optimize and consolidate code

Optimized some code by changing classes and methods
in some logics with free functions.

Consolidated the debian repo parsing.

waterflow80 committed:
Fix cachedir path formatting issue

waterflow80 committed:
fixup! Complete lzreposync service entry point

waterflow80 committed:
Complete gpg signature check for rpm

Completed the gpg signature check for rpm repositories,
mainly for the repomd.xml file.

This is done by downloading the signature file from the
remote rpm repo, and executing 'gpg verify' to verify the
repomd.xml file against its signature using the already
added gpg keys on the filesystem.

So, if you haven't already added the required gpg keyring
on your system, you'll not be able to verify the repo.

You should ideally run this version directly on the uyuni-
server, because the gpg keyring will probably be present
there.

waterflow80 committed:
fixup! Add remote_path column

agraul committed:
Refactor: Allow more input variants in makedirs()

makedirs() in uyuni.common.fileutils now accepts relative paths that
consist of only a directory name or paths with trailing slashes.

waterflow80 committed:
Complete gpg signature check for debian

Completed the gpg signature check for debian repositories.

If you haven't already added the required gpg keyring
on your system, you'll not be able to verify the repo,
and you'll normally get a GeneralRepoException.

You should ideally run this version directly on the uyuni-
server, because the gpg keyring will probably be present
there.

waterflow80 committed:
Mock spacewak gpg home directory

Mocked the SPACEWALK_GPG_HOMEDIR value to `~/.gnupg/`, which is the
default directory for gpg, in order to execute the gpg tests outside
the uyuni-server
